apps/heft/src/cli/actions/TestAction.ts

```diff
@@ -1,115 +1,87 @@
+    super(heftActionOptions, heftSession, {
       actionName: 'test',
       summary: 'Build the project and run tests.',
       documentation: ''
     });
-
-    this._buildAction = options.buildAction;
   }
 
   public onDefineParameters(): void {
     super.onDefineParameters();
 
+    this._noTestFlag = this.defineFlagParameter({
+      parameterLongName: '--notest',
+      description: 'If specified, run the build without testing.'
+    });
+
     this._noBuildFlag = this.defineFlagParameter({
       parameterLongName: '--no-build',
       description: 'If provided, only run tests. Do not build first.'
     });
   }
 
-  protected async actionExecute(buildActionContext: IBuildActionContext): Promise<void> {
-    const testActionContext: ITestActionContext = {
-      hooks: new TestHooks(),
-      properties: {
-        watchMode: buildActionContext.properties.watchMode,
-        productionFlag: buildActionContext.properties.productionFlag
-      }
-    };
+  protected async actionExecuteAsync(): Promise<void> {
     const shouldBuild: boolean = !this._noBuildFlag.value;
+    const watchMode: boolean = this._watchFlag.value;
+    const noTest: boolean = this._noTestFlag.value;
+    const lite: boolean = this._liteFlag.value;
 
-    if (testActionContext.properties.watchMode) {
+    if (watchMode) {
       if (!shouldBuild) {
         throw new Error(`${this._watchFlag.longName} is not compatible with ${this._noBuildFlag.longName}`);
-      } else if (buildActionContext.properties.noTest) {
+      } else if (noTest) {
         throw new Error(`${this._watchFlag.longName} is not compatible with ${this._noTestFlag.longName}`);
+      } else if (lite) {
+        throw new Error(`${this._watchFlag.longName} is not compatible with ${this._liteFlag.longName}`);
       }
     }
 
-    this.testActionHook.call(testActionContext);
-
-    if (testActionContext.hooks.overrideAction.isUsed()) {
-      await testActionContext.hooks.overrideAction.promise(buildActionContext.properties);
-      return;
+    if (!shouldBuild) {
+      if (noTest) {
+        throw new Error(`${this._noTestFlag.longName} is not compatible with ${this._noBuildFlag.longName}`);
+      }
     }
 
-    await testActionContext.hooks.loadActionConfiguration.promise();
-    await testActionContext.hooks.afterLoadActionConfiguration.promise();
-
-    if (testActionContext.properties.watchMode) {
-      // In --watch mode, run all configuration upfront and then kick off all stages
-      // concurrently with the expectation that the their promises will never resolve
-      // and that they will handle watching filesystem changes
-
-      this._buildAction.actionHook.call(buildActionContext);
-      await testActionContext.hooks.configureTest.promise();
-
-      await Promise.all([
-        super.actionExecute(buildActionContext),
-        this._runStageWithLogging('Test', testActionContext)
-      ]);
+    if (noTest || lite /* "&& shouldBuild" is implied */) {
```

Comment:
This is okay for now, although I wonder if eventually the action->stage selection could be entirely declarative in a config file, rather than having custom code for each CLI action.  Our multiphase design for Rush was modeled this way necessarily, and it didn't seem to be very awkward.  Fundamentally it seems like you could design a "run every stage and every task" mega-rig, and then the CLI actions would be essentially macros that enable different subsets of that.

Reply:
Interesting idea. That'd be fun to prototype.